src/marker.rs

```diff
@@ -825,25 +826,61 @@ impl<'py> Python<'py> {
         PythonVersionInfo::from_str(version_number_str).unwrap()
     }
 
-    /// Registers the object in the release pool, and tries to downcast to specific type.
+    /// Deprecated form of [`Python::checked_cast_as_bound`]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`Python::checked_cast_as` will be replaced by `Python::checked_cast_as_bound` in a future PyO3 version"
+        )
+    )]
     pub fn checked_cast_as<T>(self, obj: PyObject) -> Result<&'py T, PyDowncastError<'py>>
     where
         T: PyTypeCheck<AsRefTarget = T>,
     {
-        obj.into_ref(self).downcast()
+        self.checked_cast_as_bound::<T>(obj)
+            .map(Bound::into_gil_ref)
+            .map_err(PyDowncastError::from_downcast_into_err)
     }
 
-    /// Registers the object in the release pool, and does an unchecked downcast
-    /// to the specific type.
+    /// Registers the object in the release pool, and tries to downcast to specific type.
+    pub fn checked_cast_as_bound<T>(
+        self,
+        obj: PyObject,
+    ) -> Result<Bound<'py, T>, DowncastIntoError<'py>>
+    where
+        T: PyTypeCheck,
+    {
+        obj.into_bound(self).downcast_into()
+    }
+
+    /// Deprecated form of [`Python::cast_as_bound`]
     ///
     /// # Safety
     ///
     /// Callers must ensure that ensure that the cast is valid.
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`Python::cast_as` will be replaced by `Python::cast_as_bound` in a future PyO3 version"
+        )
+    )]
     pub unsafe fn cast_as<T>(self, obj: PyObject) -> &'py T
     where
         T: HasPyGilRef<AsRefTarget = T>,
     {
-        obj.into_ref(self).downcast_unchecked()
+        self.cast_as_bound::<T>(obj).into_gil_ref()
+    }
+
+    /// Registers the object in the release pool, and does an unchecked downcast
+    /// to the specific type.
+    ///
+    /// # Safety
+    ///
+    /// Callers must ensure that ensure that the cast is valid.
+    pub unsafe fn cast_as_bound<T>(self, obj: PyObject) -> Bound<'py, T> {
+        obj.into_bound(self).downcast_into_unchecked()
     }
```

Comment:
I think we don't need to offer `cast_as_bound` or `checked_cast_as_bound` APIs and should instead just deprecate these APIs recommending to use `Py::downcast_bound_unchecked` and `Py::downcast_bound` instead.

The advantage of `cast_as` over `Py::downcast_unchecked` was purely to place a `PyObject` into the GIL Pool and create a GIL Ref. I think this API might even be older than `Py::into_ref`.

Reply:
Great! They already felt like a little duplication to me (and to me naming felt reversed anyways). I'll drop that commit and add the deprecation warnings instead.